lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -676,22 +731,47 @@ def _on_config_changed(self, event: ConfigChangedEvent):  # noqa C901
         if not self.plugin_manager.check_plugin_manager_ready():
             return
 
+        if self.upgrade_in_progress:
+            # The following changes in _on_config_changed are not supported during an upgrade
+            # Therefore, we leave now
+            logger.warning(
+                "Changing config during an upgrade is not supported. The charm may be in a broken, "
+                "unrecoverable state"
+            )
+            event.defer()
+            return
+
         try:
             if not self.plugin_manager.check_plugin_manager_ready():
                 raise OpenSearchNotFullyReadyError()
 
             if self.unit.is_leader():
                 self.status.set(MaintenanceStatus(PluginConfigCheck), app=True)
 
-            if self.plugin_manager.run():
-                if self.upgrade_in_progress:
-                    logger.warning(
-                        "Changing config during an upgrade is not supported. The charm may be in a broken, "
-                        "unrecoverable state"
+            restart_requested = self.plugin_manager.run()
+            if (
+                self.peers_data.get(Scope.APP, PERFORMANCE_PROFILE)
+                and PerformanceType(self.peers_data.get(Scope.APP, PERFORMANCE_PROFILE))
+                != self.opensearch.perf_profile.typ
+            ):
+                self.opensearch.perf_profile = OpenSearchPerfProfile.from_str(
+                    self.config.get(PERFORMANCE_PROFILE)
+                )
+                # If we have a running service, and our profile changed
+                # then we need a restart to apply the new profile
+                self.opensearch_config.apply_performance_profile(self.opensearch.perf_profile)
+                self._apply_profile_templates_event.emit()
```

Comment:
I don't see the need of this

Reply:
Check my comment here: https://github.com/canonical/opensearch-operator/pull/466/files/51d92e7074c0b37409ea279f5f328151e244218c#r1800783632